subsys/bluetooth/host/conn.c

```diff
@@ -614,429 +592,460 @@ static inline uint16_t conn_mtu(struct bt_conn *conn)
+}
 
-	LOG_DBG("conn %p buf %p len %u flags 0x%02x", conn, buf, buf->len,
-		flags);
+static bool is_iso_conn(struct bt_conn *conn)
+{
+	return IS_ENABLED(CONFIG_BT_ISO_TX) &&
+		conn->type == BT_CONN_TYPE_ISO;
+}
 
-	/* Add to pending, it must be done before bt_buf_set_type */
-	key = irq_lock();
-	if (tx) {
-		sys_slist_append(&conn->tx_pending, &tx->node);
-	} else {
-		struct bt_conn_tx *tail_tx;
+static bool is_le_conn(struct bt_conn *conn)
+{
+	return IS_ENABLED(CONFIG_BT_CONN) && conn->type == BT_CONN_TYPE_LE;
+}
 
-		tail_tx = (void *)sys_slist_peek_tail(&conn->tx_pending);
-		if (tail_tx) {
-			pending_no_cb = &tail_tx->pending_no_cb;
-		} else {
-			pending_no_cb = &conn->pending_no_cb;
-		}
+static bool is_acl_conn(struct bt_conn *conn)
+{
+	return is_le_conn(conn) || is_classic_conn(conn);
+}
 
-		(*pending_no_cb)++;
+static int send_buf(struct bt_conn *conn, struct net_buf *buf,
+		    void *cb, void *ud)
+{
+	struct net_buf *frag = NULL;
+	struct bt_conn_tx *tx = NULL;
+	uint8_t flags;
+	int err;
+
+	if (buf->len == 0) {
+		__ASSERT_NO_MSG(0);
+
+		return -EMSGSIZE;
 	}
-	irq_unlock(key);
 
-	if (IS_ENABLED(CONFIG_BT_ISO) && conn->type == BT_CONN_TYPE_ISO) {
-		err = send_iso(conn, buf, flags);
-	} else if (IS_ENABLED(CONFIG_BT_CONN)) {
-		err = send_acl(conn, buf, flags);
-	} else {
-		__ASSERT(false, "Invalid connection type %u", conn->type);
+	if (bt_buf_has_view(buf)) {
+		__ASSERT_NO_MSG(0);
+
+		return -EIO;
 	}
 
-	if (err) {
-		LOG_ERR("Unable to send to driver (err %d)", err);
-		key = irq_lock();
-		/* Roll back the pending TX info */
-		if (tx) {
-			sys_slist_find_and_remove(&conn->tx_pending, &tx->node);
-		} else {
-			__ASSERT_NO_MSG(*pending_no_cb > 0);
-			(*pending_no_cb)--;
-		}
-		irq_unlock(key);
+	LOG_DBG("conn %p buf %p len %u", conn, buf, buf->len);
 
-		/* We don't want to end up in a situation where send_acl/iso
-		 * returns the same error code as when we don't get a buffer in
-		 * time.
+	/* Acquire the right to send 1 packet to the controller */
+	if (k_sem_take(bt_conn_get_pkts(conn), K_NO_WAIT)) {
+		/* This shouldn't happen now that we acquire the resources
+		 * before calling `send_buf` (in `get_conn_ready`). We say
+		 * "acquire" as `tx_processor()` is not re-entrant and the
+		 * thread is non-preemptible. So the sem value shouldn't change.
 		 */
-		err = -EIO;
-		goto fail;
+		__ASSERT(0, "No controller bufs");
+
+		return -ENOMEM;
 	}
 
-	return 0;
+	/* Allocate and set the TX context */
+	tx = conn_tx_alloc();
 
-fail:
-	/* If we get here, something has seriously gone wrong:
-	 * We also need to destroy the `parent` buf.
-	 */
-	k_sem_give(bt_conn_get_pkts(conn));
-	if (tx) {
-		/* `buf` might not get destroyed, and its `tx` pointer will still be reachable.
-		 * Make sure that we don't try to use the destroyed context later.
-		 */
-		tx_data(buf)->tx = NULL;
-		conn_tx_destroy(conn, tx);
+	/* See big comment above */
+	if (!tx) {
+		__ASSERT(0, "No TX context");
+
+		return -ENOMEM;
```

Comment:
Do we need to return the semaphore here in case of `tx == NULL`?

Reply:
What about later error handling in this function?

Reply:
sooo. This function without asserts-on is essentially broken. It makes little sense to do anything with Bluetooth after this, as the host is completely out of sync with the controller. That's the kinda thing why asserts should _stay_ on.

In my opinion, such "error handling" is misguided. It litters the code, adding mental overhead to readers/reviewers for no gain, only to satisfy people that want to run code incorrectly.

I'm now convinced and will make the PR to force asserts on; made https://github.com/zephyrproject-rtos/zephyr/issues/73316